bfgs.go

```diff
@@ -128,34 +121,31 @@ func (b *BFGS) NextDirection(loc *Location, dir []float64) (stepSize float64) {
 		b.first = false
 	}
 
-	// Compute the update rule
-	//     B_{k+1}^-1
-	// First term is just the existing inverse hessian
-	// Second term is
-	//     (sk^T yk + yk^T B_k^-1 yk)(s_k sk_^T) / (sk^T yk)^2
-	// Third term is
-	//     B_k ^-1 y_k sk^T + s_k y_k^T B_k-1
-	//
-	// y_k^T B_k^-1 y_k is a scalar, and the third term is a rank-two update
-	// where B_k^-1 y_k is one vector and s_k is the other. Compute the update
-	// values then actually perform the rank updates.
-	yBy := mat64.Inner(b.yVec, b.invHess, b.yVec)
-	firstTermConst := (sDotY + yBy) / (sDotYSquared)
-	b.tmpVec.MulVec(b.invHess, b.yVec)
-
-	b.invHess.RankTwo(b.invHess, -1/sDotY, b.tmpVec, b.sVec)
-	b.invHess.SymRankOne(b.invHess, firstTermConst, b.sVec)
-
-	// update the bfgs stored data to the new iteration
-	copy(b.x, loc.X)
-	copy(b.grad, loc.Gradient)
-
-	// Compute the new search direction
-	d := mat64.NewVector(b.dim, dir)
-	g := mat64.NewVector(b.dim, loc.Gradient)
-
-	d.MulVec(b.invHess, g) // new direction stored in place
-	floats.Scale(-1, dir)
+	if math.Abs(sDotY) != 0 {
```

Comment:
Good question. I have seen the test `if sDotY > 1e-14 { update }` used elsewhere. If I plug in the same check, several test fail. I decrease the constant, the number of failed tests also decreases, but some obviously take much longer to finish. The smallest I tried was 1e-18, then all tests pass.

Since that constant is kind of heuristical and arbitrary, for the moment I would leave the test as is. It fixes the problem I had and all tests quickly pass.